node/core/av-store/src/lib.rs

```
@@ -0,0 +1,555 @@
+struct StoredAvailableData {
+	data: AvailableData,
+	n_validators: u32,
+}
+
+/// Configuration for the availability store.
+pub struct Config {
+	/// Cache size in bytes. If `None` default is used.
+	pub cache_size: Option<usize>,
+	/// Path to the database.
+	pub path: PathBuf,
+}
+
+impl AvailabilityStoreSubsystem {
+	/// Create a new `AvailabilityStoreSubsystem` with a given config on disk.
+	pub fn new_on_disk(config: Config) -> io::Result<Self> {
+		let mut db_config = DatabaseConfig::with_columns(columns::NUM_COLUMNS);
+
+		if let Some(cache_size) = config.cache_size {
+			let mut memory_budget = HashMap::new();
+
+			for i in 0..columns::NUM_COLUMNS {
+				memory_budget.insert(i, cache_size / columns::NUM_COLUMNS as usize);
+			}
+			db_config.memory_budget = memory_budget;
+		}
+
+		let path = config.path.to_str().ok_or_else(|| io::Error::new(
+			io::ErrorKind::Other,
+			format!("Bad database path: {:?}", config.path),
+		))?;
+
+		let db = Database::open(&db_config, &path)?;
+
+		Ok(Self {
+			inner: Arc::new(db),
+		})
+	}
+
+	#[cfg(test)]
+	fn new_in_memory(inner: Arc<dyn KeyValueDB>) -> Self {
+		Self {
+			inner,
+		}
+	}
+}
+
+async fn run<Context>(subsystem: AvailabilityStoreSubsystem, mut ctx: Context)
+	-> Result<(), Error>
+where
+	Context: SubsystemContext<Message=AvailabilityStoreMessage>,
+{
+	let ctx = &mut ctx;
+	loop {
+		select! {
+			incoming = ctx.recv().fuse() => {
+				match incoming {
+					Ok(FromOverseer::Signal(Conclude)) => break,
+					Ok(FromOverseer::Signal(_)) => (),
+					Ok(FromOverseer::Communication { msg }) => {
+						let _ = process_message(&subsystem.inner, msg);
+					}
+					Err(_) => break,
+				}
+			}
+			complete => break,
+		}
+	}
+
+	Ok(())
+}
+
+fn process_message(db: &Arc<dyn KeyValueDB>, msg: AvailabilityStoreMessage) -> Result<(), Error> {
+	use AvailabilityStoreMessage::*;
+	match msg {
+		QueryAvailableData(hash, tx) => {
+			let _ = tx.send(available_data(db, &hash).map(|d| d.data));
```

Is it safe to swallow send errors? Why not just propagate them?